### ISSUE TITLE
[BUG]: Yarn install fails when using Cache@2 and Yarn >1

### ISSUE THREAD

**Issue by dombarnes** [microsoft/azure-pipelines-tasks#19335]
### New issue checklist

- [X] I searched for [existing GitHub issues](https://github.com/microsoft/azure-pipelines-tasks/issues)
- [X] I read [pipeline troubleshooting guide](https://docs.microsoft.com/vsts/build-release/actions/troubleshooting)
- [X] I checked [how to collect logs](https://learn.microsoft.com/azure/devops/pipelines/troubleshooting/review-logs?view=azure-devops)

### Task name

Cache@2

### Task version

2.198.0

### Issue Description

When using Yarn >1, an error occurs after using Cace@2. Cache sems to set some config variable that is invalid.
> Usage Error: Unrecognized or legacy configuration settings found: cacheSuccess - run "yarn config -v" to see the list of settings supported in Yarn (in <environment>)

### Environment type (Please select at least one enviroment where you face this issue)

- [ ] Self-Hosted
- [X] Microsoft Hosted
- [ ] VMSS Pool
- [ ] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

windows-latest

### Relevant log output

```shell
2023-12-02T09:13:23.2350003Z ##[section]Starting: Cache Yarn packages
2023-12-02T09:13:23.2458241Z ==============================================================================
2023-12-02T09:13:23.2458690Z Task         : Cache
2023-12-02T09:13:23.2458792Z Description  : Cache files between runs
2023-12-02T09:13:23.2458994Z Version      : 2.198.0
2023-12-02T09:13:23.2459122Z Author       : Microsoft Corporation
2023-12-02T09:13:23.2459307Z Help         : https://aka.ms/pipeline-caching-docs
2023-12-02T09:13:23.2459553Z ==============================================================================
2023-12-02T09:13:24.4451818Z Resolving key:
2023-12-02T09:13:24.4616381Z  - yarn                                               [string]
2023-12-02T09:13:24.4620466Z  - "Windows_NT"                                       [string]
2023-12-02T09:13:24.5061259Z  - D:\a\1\s/<my project>/ya...                        [file] --> C50AE5F0E0BA2C48E97CC5C0D30E842BFAD91DC1DC7EB8661B3D74EAFA05DD73
2023-12-02T09:13:24.5105011Z Resolved to: yarn|"Windows_NT"|8HJwMp/Cfb2FVqsj7LpEVXRwpU7A4K5tJJMd/a28NVU=
2023-12-02T09:13:24.5163488Z Resolving restore key:
2023-12-02T09:13:24.5164044Z  - yarn         [string]
2023-12-02T09:13:24.5164941Z  - "Windows_NT" [string]
2023-12-02T09:13:24.5170057Z Resolved to: yarn|"Windows_NT"|**
2023-12-02T09:13:24.5170452Z Resolving restore key:
2023-12-02T09:13:24.5170795Z  - yarn [string]
2023-12-02T09:13:24.5171038Z Resolved to: yarn|**
2023-12-02T09:13:24.5623015Z Using default max parallelism.
2023-12-02T09:13:25.2841182Z Using default max parallelism.
2023-12-02T09:13:25.2849808Z Max dedup parallelism: 192
2023-12-02T09:13:25.2850319Z DomainId: 0
2023-12-02T09:13:25.4220691Z ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session 80983fa5-97bb-483b-84fa-560d62a96834
2023-12-02T09:13:25.4240446Z Hashtype: Dedup64K
2023-12-02T09:13:25.6356316Z Getting a pipeline cache artifact with one of the following fingerprints:
2023-12-02T09:13:25.6357575Z Fingerprint: `yarn|"Windows_NT"|8HJwMp/Cfb2FVqsj7LpEVXRwpU7A4K5tJJMd/a28NVU=`
2023-12-02T09:13:25.6358401Z Fingerprint: `yarn|"Windows_NT"|**`
2023-12-02T09:13:25.6358874Z Fingerprint: `yarn|**`
2023-12-02T09:13:25.9929182Z There is a cache miss.
2023-12-02T09:13:26.0887543Z ApplicationInsightsTelemetrySender correlated 1 events with X-TFS-Session 80983fa5-97bb-483b-84fa-560d62a96834
2023-12-02T09:13:26.1464410Z ##[section]Finishing: Cache Yarn packages

2023-12-02T09:13:29.3314660Z ##[section]Starting: Enable Corepack and setup yarn
2023-12-02T09:13:29.3423381Z ==============================================================================
2023-12-02T09:13:29.3423718Z Task         : Bash
2023-12-02T09:13:29.3423815Z Description  : Run a Bash script on macOS, Linux, or Windows
2023-12-02T09:13:29.3424137Z Version      : 3.231.0
2023-12-02T09:13:29.3424260Z Author       : Microsoft Corporation
2023-12-02T09:13:29.3424446Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2023-12-02T09:13:29.3424854Z ==============================================================================
2023-12-02T09:13:31.3034502Z Generating script.
2023-12-02T09:13:31.3038183Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-02T09:13:31.3038553Z /d/a/_temp
2023-12-02T09:13:31.3040980Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-02T09:13:31.3041379Z /d/a
2023-12-02T09:13:31.3042051Z ========================== Starting Command Output ===========================
2023-12-02T09:13:31.3044201Z [command]"C:\Program Files\Git\bin\bash.exe" /d/a/_temp/69afad6b-4409-4e11-ba6a-121fd283aa9b.sh
2023-12-02T09:13:31.3044797Z Enabling Corepack
2023-12-02T09:13:37.0183407Z Requesting Yarn version 3.6.4
2023-12-02T09:13:38.4464995Z [31m[1mUsage Error[22m[39m: Unrecognized or legacy configuration settings found: cacheSuccess - run "yarn config -v" to see the list of settings supported in Yarn (in <environment>)
2023-12-02T09:13:38.4466748Z 
2023-12-02T09:13:38.4467361Z [1m$ [22myarn set version [--only-if-needed] <version>
2023-12-02T09:13:39.2530657Z Running Yarn 3.6.4
2023-12-02T09:13:39.2640489Z ##[section]Finishing: Enable Corepack and setup yarn

2023-12-02T09:13:39.2705350Z ##[section]Starting: Install dependencies
2023-12-02T09:13:39.2818823Z ==============================================================================
2023-12-02T09:13:39.2819190Z Task         : Bash
2023-12-02T09:13:39.2819298Z Description  : Run a Bash script on macOS, Linux, or Windows
2023-12-02T09:13:39.2819609Z Version      : 3.231.0
2023-12-02T09:13:39.2819744Z Author       : Microsoft Corporation
2023-12-02T09:13:39.2819941Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2023-12-02T09:13:39.2821148Z ==============================================================================
2023-12-02T09:13:39.4777641Z Generating script.
2023-12-02T09:13:39.4956279Z Script contents:
2023-12-02T09:13:39.4957256Z yarn install --immutable
2023-12-02T09:13:39.4962023Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-02T09:13:39.5354216Z /d/a/_temp
2023-12-02T09:13:39.5404186Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-02T09:13:39.5778201Z /d/a
2023-12-02T09:13:39.5883046Z ========================== Starting Command Output ===========================
2023-12-02T09:13:39.5898349Z [command]"C:\Program Files\Git\bin\bash.exe" /d/a/_temp/4a89c342-1007-4b04-b4aa-b3d3489c12e0.sh
2023-12-02T09:13:40.4081785Z [31m[1mUsage Error[22m[39m: Unrecognized or legacy configuration settings found: cacheSuccess - run "yarn config -v" to see the list of settings supported in Yarn (in <environment>)
2023-12-02T09:13:40.4083516Z 
2023-12-02T09:13:40.4084547Z [1m$ [22myarn install [--json] [--immutable] [--immutable-cache] [--check-cache] [--inline-builds] [--mode #0]
2023-12-02T09:13:40.4475052Z ##[error]Bash exited with code '1'.
2023-12-02T09:13:40.4516507Z ##[section]Finishing: Install dependencies
```

### Full task logs with system.debug enabled

<details>
  <pre> [REPLACE THIS WITH YOUR INFORMATION] </pre>
</details>

### Repro steps

```yml
variables:
- name: YARN_CACHE_FOLDER
  value: '.yarn/'
- name: YARN_VERSION
  value: '3.6.4'
- name: YARN_CACHE_SUCCESS
  value: false
steps:
  - task: Cache@2
    displayName: Cache Yarn packages
    inputs:
      key: 'yarn | "$(Agent.OS)" | $(System.DefaultWorkingDirectory)/TrilbyTV/Trilby.TrilbyTV.Web.Portal/yarn.lock'
      restoreKeys: |
        yarn | "$(Agent.OS)"
        yarn
      path: $(System.DefaultWorkingDirectory)/MyProject/$(YARN_CACHE_FOLDER)
  - task: NodeTool@0
    inputs:
      versionSpec: '18.x'
    displayName: 'Install Node.js 18'
  - bash: |
      echo "Enabling Corepack"
      corepack enable
      echo "Requesting Yarn version $(YARN_VERSION)"
      yarn set version $(YARN_VERSION)
      echo "Running Yarn $(yarn --version)"
    displayName: 'Enable Corepack and setup yarn'
    workingDirectory: '$(System.DefaultWorkingDirectory)/MyProject/'
  - bash: |
      yarn install
    displayName: 'Install dependencies'
    condition: ne(variables.YARN_CACHE_SUCCESS, true)
```

**Comment by dombarnes** [comment]
After further debugging I found YARN_CACHE_SUCCESS=false is already set in the environment. Oddly this does not happen in all my jobs where I perform this. 

’‘‘
2023-12-03T01:21:02.8702052Z ##[section]Starting: Enable Corepack and setup yarn
2023-12-03T01:21:02.8833537Z ==============================================================================
2023-12-03T01:21:02.8833940Z Task         : Bash
2023-12-03T01:21:02.8834053Z Description  : Run a Bash script on macOS, Linux, or Windows
2023-12-03T01:21:02.8834425Z Version      : 3.231.0
2023-12-03T01:21:02.8834574Z Author       : Microsoft Corporation
2023-12-03T01:21:02.8834796Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2023-12-03T01:21:02.8835283Z ==============================================================================
2023-12-03T01:21:03.1283205Z Generating script.
2023-12-03T01:21:03.1526762Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-03T01:21:04.0023634Z /d/a/_temp
2023-12-03T01:21:05.4796808Z [command]"C:\Program Files\Git\bin\bash.exe" -c pwd
2023-12-03T01:21:05.4797764Z /d/a
2023-12-03T01:21:05.4798676Z ========================== Starting Command Output ===========================
2023-12-03T01:21:05.4801378Z [command]"C:\Program Files\Git\bin\bash.exe" /d/a/_temp/634108b0-6d63-4e36-bf5d-40bab8844a03.sh
2023-12-03T01:21:05.4802084Z Enabling Corepack
2023-12-03T01:21:10.1595662Z Requesting Yarn version 3.6.4
2023-12-03T01:21:12.0481084Z ➤ YN0000: Retrieving https://repo.yarnpkg.com/3.6.4/packages/yarnpkg-cli/bin/yarn.js
2023-12-03T01:21:12.7726725Z ➤ YN0000: Saving the new release in .yarn/releases/yarn-3.6.4.cjs
2023-12-03T01:21:12.7783461Z ➤ YN0000: Done in 0s 742ms
2023-12-03T01:21:13.7099750Z Running Yarn 3.6.4
2023-12-03T01:21:13.7273610Z ##[section]Finishing: Enable Corepack and setup yarn
’‘‘

**Comment by dombarnes** [comment]
Figured this out i think. I had used YARN_CACHE_SUCCESS for the cacheHitVar input option under a Cache@2 task, and was defining this in my variables. It seems Yarn parses any YARN_ prefixed environment variable and then fails on it. So I've had to change my variable name to something else.
But undocumented but easy to trip up on this.
Leaving this issue open in case there's recommended documentation updates.
